### DISCUSSION TITLE
Why does the with-operator need parenthesis around the expression to access object.

### DISCUSSION THREAD

**Discussion by FreeApophis** [dotnet/csharplang#5378] [Q&A]
We found very strange behaviour on the `with`-operator. I'd like to discuss if this is an oversight or if there are some real implications.

When you create an instance of Type T with the Property Initializer, the expression itself is of type T.
However if you create an instance of Type T using a with-expression Initializer, I cannot access the Interface of the Type T.

Here is a full example:

```cs
public record ComplexNumber
{
    public double Re { get; init; }
    public double Im { get; init; }
    public double R => Math.Sqrt((Re * Re) + (Im * Im));
    public double Phi => Math.Atan2(Im, Re);
}

var zero= new ComplexNumber { Re = 4 };
var fromInitializer = new ComplexNumber { Im = 3 }.R;
var fromWithExpression = zero with { Im = 3 }.R; // this does not compile!
var working = (zero with { Im = 3 }).R; // With parenthesis, it works as expected.
```

When analyzing the syntax-tree with Rosly-Quoter, it looks like the Syntax-Tree is very similar, and I think it should be handled similarly. 

Could you please point me to the C# language specification which describes this behaviour, I have read [the 5 paragraphs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records#with-expression) about the with-expression, I did not find an explanation.

**Comment by HaloFour** [comment]
I'm going to suggest that it's possible to ask this question about being required to wrap the "with" expression in parenthesis without being combative.

**Reply by FreeApophis** [reply]
Ok, could you please point me to the specification which describes this behaviour?

**Reply by DavidArno** [reply]
Have a look at the [Operator precedence on the C# operators and expressions (C# reference) page](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/#operator-precedence). The member access expression (`x.y`) has higher precedence than the `with` expression. So it's the `{ Im = 3 }.R` that isn't compiling as the compiler sees it as `(zero) with ({ Im = 3 }.R)`.

**Reply by Mafii** [reply]
Is there any reason the brackets after `with` bind stronger to the right side than to the with? As that part doesn't have any meaning in itself without the `with` and the variable before it, it was very confusing to read that the compiler has this precedence. Wouldn't this be something that could be changed without any meaningful implications?

**Reply by FreeApophis** [reply]
I actually read the precedence list, and obviously you want that the `.` has a higher precedence in a case like this:

```cs
object.Property with { Value = 42 }`
```

However, I do not see why the Initilizer list couldn't be part of the with-operator, like the intializer list is part of the constructor.

**Comment by svick** [comment]
This discussion about choosing the precedence of the `switch` expression may be of interest: https://github.com/dotnet/csharpstandard/issues/300.

[Since the `with` expression has the same precedence as the `switch` expression](https://github.com/dotnet/roslyn/blob/30c493a/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs#L10228-L10230), and since they have similar syntaxes, I think the same considerations should apply to both.

**Reply by DavidArno** [reply]
That's interesting as [the docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/#operator-precedence) suggest `switch` has a higher precedence than `with`.

**Reply by agocke** [reply]
@svick is basically right here -- the precedence of `switch` and `with` were intended to be the same, just because they look and act so similar. Beyond that, any behaviors for `with` were inherited from `switch`. I'm not sure why the docs have them listed differently, I think that was just a mistake during transcription

**Reply by svick** [reply]
I have opened a PR to fix the docs: https://github.com/dotnet/docs/pull/26789.

**Reply by DavidArno** [reply]
@svick, you beat me to it, was about to do that myself. I'll put my feet up instead 😃

**Reply by FreeApophis** [reply]
Thanks, that is very interesting, and I added to the dicussion there.
